packages/backend-common/src/middleware/authenticationHandler.ts

```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright 2020 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { NextFunction, Request, RequestHandler, Response } from 'express';
+
+export interface Authenticator {
+  authenticateRequest: (request: Request) => boolean;
+
+  // log the failed request if needed
+  logRequest?: (request: Request) => void;
+}
+
+/**
+ * Express middleware to authenticate requests
+ *
+ * @returns An Express request handler
+ */
+export function authenticationHandler(
```

Comment:
This ends up being a bit of a middleware then then runs another middleware? ^_>

I think a middleware that is a bit more direct in handling is Backstage ID Tokens would make sense. Not exactly sure what we want the contract to be, but I would imagine something like `IdToken -> UserEntity`. I think we could allow for other IdTokens to be plugged in as well, as long as there is a matching configuration of the `SignInPage` in the frontend. Either way we want to support IdTokens issued by the `auth-backend` so that we can support non-OIDC or multi-provider auth flows.

Reply:
I think , perhaps that can be solved by a utility class that implements the `Authenticator` interface ? 
As you mentioned, the Idtokens can either be generated by backstage via the `TokenIssuer` interface or can be provided by the Auth provider. 

[TokenFactor]( https://github.com/backstage/backstage/blob/a0a3e6d6cacdeb2218065c6bacf6be1485bff175/plugins/auth-backend/src/identity/TokenFactory.ts) has an interface to generate a Token, an optional method can be added to it to verify a token. 

Similarly all the Wrappers for `OAuth` and other Auth providers can additionally expose a method to return an `authenticator` instance that can authenticate tokens issued by them. 
When wrapping routers with this middleware, Applications can explicitly mention the `provider` whose authenticator is used to wrap certain routes. 

I kept the middleware style so as to also allow other middlewares that are in the chain to be processed . A single `Router` instance might have possible many middleware in the chain.